src/resource/normal.d.ts

```diff
@@ -25,25 +35,41 @@ declare namespace schemas {
     _processStrategy: StrategyFunction<T>;
   }
 
-  export class Object<T = any> {
-    constructor(definition: { [key: string]: Schema<T> });
+  export class Object<
+    T = any,
+    O extends { [key: string]: Schema<T> } = { [key: string]: Schema<T> }
+  > {
+    constructor(definition: O);
     define(definition: Schema): void;
   }
 
-  export class Union<T = any> {
+  export class Union<Choices extends EntityMap = any> {
     constructor(
-      definition: Schema<T>,
-      schemaAttribute?: string | SchemaFunction,
+      definition: Choices,
+      schemaAttribute:
+        | (Choices[keyof Choices] extends Entity<infer T> ? keyof T : never)
+        | SchemaFunction<keyof Choices>,
     );
     define(definition: Schema): void;
+    inferSchema: SchemaAttributeFunction<Choices[keyof Choices]>;
+    getSchemaAttribute: SchemaFunction<keyof Choices>;
   }
 
-  export class Values<T = any> {
+  export class Values<Choices extends EntityMap | Schema = any> {
```

Comment:
values have these two forms (https://github.com/paularmstrong/normalizr/blob/master/docs/api.md#valuesdefinition-schemaattribute) so we end up with many ternary type operators when using it to distinguish between the two.